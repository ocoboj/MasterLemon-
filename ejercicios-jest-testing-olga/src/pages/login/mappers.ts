import * as model from '../../rest-api/model';
import * as vm from './viewModel';

export const mapLoginEntityVMToModel = (loginEntity: vm.LoginEntity): model.LoginEntity => (
  Boolean(loginEntity) ?
  {
    ...loginEntity,
  } :
  null
);
