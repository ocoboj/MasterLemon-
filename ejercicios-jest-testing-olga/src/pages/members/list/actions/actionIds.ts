export const actionIds = {
  UPDATE_MEMBERS: 'UPDATE_MEMBERS',
};
