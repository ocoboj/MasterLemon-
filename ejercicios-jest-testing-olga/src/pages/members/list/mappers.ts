import * as model from '../../../rest-api/model';
import * as vm from './viewModel';

export const mapMemberListModelToVM = (members: model.Member[]): vm.Member[] => (
  Array.isArray(members) ?
    members.map(mapMemberModelToVM) :
    []
);

const mapMemberModelToVM = (member: model.Member): vm.Member => ({
  id: member.id,
  name: member.login,
  avatarUrl: member.avatar_url,
});
