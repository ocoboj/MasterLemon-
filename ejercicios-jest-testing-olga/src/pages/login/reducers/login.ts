import { actionIds } from '../actions/actionIds';
import {
  LoginEntity, createEmptyLoginEntity,
  LoginFormErrors, createEmptyLoginFormErrors,
} from '../viewModel';

export interface LoginState {
  loginEntity: LoginEntity;
  loginFormErrors: LoginFormErrors;
}

const createEmptyState = (): LoginState => ({
  loginEntity: createEmptyLoginEntity(),
  loginFormErrors: createEmptyLoginFormErrors(),
});

export const loginReducer = (state = createEmptyState(), action) => {
  switch (action.type) {
    case actionIds.UPDATE_LOGIN_ENTITY_FIELD:
      return handleUpdateLoginEntityField(state, action.payload);

    case actionIds.UPDATE_LOGIN_FORM_ERRORS:
      return handleUpdateLoginFormErrors(state, action.payload);
  }

  return state;
};

const handleUpdateLoginEntityField = (state: LoginState, { fieldName, value, fieldValidationResult }): LoginState => ({
  loginEntity: {
    ...state.loginEntity,
    [fieldName]: value,
  },
  loginFormErrors: {
    ...state.loginFormErrors,
    [fieldName]: fieldValidationResult,
  },
});

const handleUpdateLoginFormErrors = (state: LoginState, loginFormErrors: LoginFormErrors): LoginState => ({
  ...state,
  loginFormErrors,
});
