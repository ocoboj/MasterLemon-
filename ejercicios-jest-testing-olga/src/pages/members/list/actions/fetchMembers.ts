import { actionIds } from './actionIds';
import { Member } from '../../../../rest-api/model';
import * as apiMember from '../../../../rest-api/api/member';

export const fetchMembers = () => (dispatch) => (
  apiMember.fetchMembers()
    .then((members) => dispatch(fetchMembersCompleted(members)))
    .catch(console.log)
);

const fetchMembersCompleted = (members: Member[]) => ({
  type: actionIds.UPDATE_MEMBERS,
  payload: members,
});
