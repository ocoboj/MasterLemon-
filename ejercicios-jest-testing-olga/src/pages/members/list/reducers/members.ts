import { actionIds } from '../actions/actionIds';
import { Member } from '../../../../rest-api/model';

export type MembersState = Member[];

export const membersReducer = (state = [], action): MembersState => {
  switch (action.type) {
    case actionIds.UPDATE_MEMBERS:
      return handleUpdateMembers(state, action.payload);
  }

  return state;
};

const handleUpdateMembers = (state: MembersState, members: Member[]): MembersState => (
  members
);
