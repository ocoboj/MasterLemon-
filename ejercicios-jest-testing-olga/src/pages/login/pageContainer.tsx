import { connect } from 'react-redux';
import { State } from '../reducers';
import { updateLoginEntityField } from './actions/updateLoginEntityField';
import { loginRequest } from './actions/loginRequest';
import { LoginEntity } from './viewModel';
import { LoginPage } from './page';

const mapStateToProps = (state: State) => ({
  loginEntity: state.login.loginEntity,
  loginFormErrors: state.login.loginFormErrors,
});

const mapDispatchToProps = (dispatch) => ({
  updateField: (loginEntity: LoginEntity) => (fieldName: string, value: any) => dispatch(
    updateLoginEntityField(loginEntity, fieldName, value)
  ),
  doLogin: (loginEntity: LoginEntity) => () => dispatch(loginRequest(loginEntity)),
});

const mergeProps = (stateProps, dispatchProps, ownProps) => ({
  ...ownProps,
  ...stateProps,
  updateField: dispatchProps.updateField(stateProps.loginEntity),
  doLogin: dispatchProps.doLogin(stateProps.loginEntity),
});

export const LoginPageContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
  mergeProps
)(LoginPage);
