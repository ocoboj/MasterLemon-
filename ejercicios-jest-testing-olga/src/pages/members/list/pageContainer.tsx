import * as React from 'react';
import { connect } from 'react-redux';
import { State } from '../../reducers';
import { fetchMembers } from './actions/fetchMembers';
import { Member } from './viewModel';
import { MemberListPage } from './page';
import { getMembersVM } from './selectors';

const mapStateToProps = (state: State) => ({
  members: getMembersVM(state),
});

const mapDispatchToProps = (dispatch) => ({
  fetchMembers: () => dispatch(fetchMembers()),
});

interface Props {
  members: Member[];
  fetchMembers: () => void;
}

class PageContainer extends React.PureComponent<Props, {}> {
  componentDidMount() {
    this.props.fetchMembers();
  }

  render() {
    return (
      <MemberListPage
        members={this.props.members}
      />
    );
  }
}

export const MemberListPageContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(PageContainer);
