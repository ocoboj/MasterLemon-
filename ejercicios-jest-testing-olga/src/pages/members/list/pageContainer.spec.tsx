import * as React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { State } from '../../reducers';
import * as mappers from './mappers';
import * as fetchMemberActions from './actions/fetchMembers';
import { MemberListPageContainer } from './pageContainer';
import { login } from '../../../rest-api/api/login';


const getMockStore = configureStore();

describe('pages/members/list/pageContainer spec', () => {

    it('should call to fetchMember action when render component', () => {
        const state = {
            members: [
                {
                    id: 4,
                    login: 'test login 4',
                    avatar_url: 'test avatar_url 4'
                },
            ],
        } as State;

        const store = getMockStore(state);

        const fetchMembers = jest.spyOn(fetchMemberActions, 'fetchMembers').mockImplementation(() => ({
            type: 'test action type',
        }));

        const component = mount(
            <MemberListPageContainer />,
            {
                context: { store },
            }
        );

        expect(fetchMembers).toHaveBeenCalled();
        expect(fetchMembers).toMatchSnapshot();
    });

    it ('should call to mapMemebersListModelToVM when render component', () => {
        const state = {
            members: [
                {
                    id: 2,
                    login: 'test login 2',
                    avatar_url: 'test avatar_url 2'

                }
            ]
        } as State;

        const store = getMockStore(state);

        const mapMemeber = jest.spyOn(mappers, 'mapMemberListModelToVM').mockReturnValue([
            {
                id: 3,
                name: 'test login 3',
                avatarUrl: 'test avatar_url 3'
            }
        ]);

        const component = shallow(
            <MemberListPageContainer />,
            {
                context: {store}
            },
        );
        expect(mapMemeber).toHaveBeenCalled();
        expect(mapMemeber).toMatchSnapshot();
    });

    it('should render when passing of the sate', () => {
        const state = {
            members: [
                {
                    id: 6,
                    login: 'test login 6',
                    avatar_url: 'test avatar_url 6'
                }
            ]
        } as State;
        
        const store = getMockStore(state);

        const component = shallow (
            <MemberListPageContainer />,
            {
                context: {store}
            }
        );
        expect(component).toMatchSnapshot();
    })
})
