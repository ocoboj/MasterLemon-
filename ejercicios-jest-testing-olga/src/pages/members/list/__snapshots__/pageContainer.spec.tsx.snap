// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/members/list/pageContainer spec should call to fetchMember action when render component 1`] = `
[MockFunction] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "type": "test action type",
      },
    },
  ],
}
`;

exports[`pages/members/list/pageContainer spec should call to mapMemebersListModelToVM when render component 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "avatar_url": "test avatar_url 2",
          "id": 2,
          "login": "test login 2",
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Array [
        Object {
          "avatarUrl": "test avatar_url 3",
          "id": 3,
          "name": "test login 3",
        },
      ],
    },
  ],
}
`;

exports[`pages/members/list/pageContainer spec should render when passing of the sate 1`] = `
<PageContainer
  fetchMembers={[Function]}
  members={
    Array [
      Object {
        "avatarUrl": "test avatar_url 6",
        "id": 6,
        "name": "test login 6",
      },
    ]
  }
/>
`;
